src/import/import.rs

```diff
@@ -34,23 +34,110 @@ use super::{Config, Error, Result};
 const MAX_RETRY_TIMES: u64 = 5;
 const RETRY_INTERVAL_SECS: u64 = 3;
 
+/// Speed limit enforcement.
+///
+/// We need to restrict the maximum upload speed to avoid saturating the network
+/// bandwidth, which caused tikv-importer unable to contact PD.
+///
+/// This is implemented using the Token Bucket algorithm based on [1], but
+/// allowing bursting.
+///
+/// [1]: https://medium.com/smyte/rate-limiter-df3408325846
+struct SpeedLimiter {
+    // The maximum speed limit, unit = Byte/s.
+    speed_limit: f64,
+    // The last update time and the remaining amount of bytes can be sent during
+    // this second.
+    value: Mutex<(Instant, f64)>,
+}
+
+impl SpeedLimiter {
+    /// Creates a speed limiter.
+    fn new(cfg: &Config) -> Self {
+        let speed_limit = cfg.upload_speed_limit.0 as f64;
+        Self {
+            speed_limit,
+            value: Mutex::new((Instant::now_coarse(), speed_limit)),
+        }
+    }
+
+    /// Takes a block of the given size (in bytes) from the speed limiter. If
+    /// the block is taken too quickly, this method will sleep until the speed
+    /// limit is below the specified value.
+    ///
+    /// This method will block the entire thread if the speed limit is violated,
+    /// and thus should not be invoked when implementing an async fn.
+    fn take(&self, tag: &str, size: u64) {
+        let z = size as f64;
+
+        loop {
+            let mut lock = self.value.lock().expect("no poison");
+            let (last_update, mut value) = *lock;
+            let now = Instant::now_coarse();
+            let elapsed = duration_to_sec(now.duration_since(last_update));
+            value = self.speed_limit.min(value + elapsed * self.speed_limit);
+            if value >= z || value >= self.speed_limit {
+                // The second condition allows take() to proceed on a completely
+                // filled bucket, i.e. allows sending a 1 GB file on a 512 MB/s
+                // limit. The bucket value will temporarily drop to -512 MB,
+                // which will be refilled after 2 seconds as expected.
+                *lock = (now, value - z);
+                return;
+            }
+
+            // Reaching here means the bucket's value is not high enough to send
+            // all bytes. We sleep for the time needed to completely refill the
+            // bucket.
+            drop(lock);
```

Comment:
Is it possible another thread acquire the lock and take the permit when this thread is sleep?

Reply:
Yes. When this happens, the other thread will get an even smaller `value`, which causes this other thread to also sleep.